packages/playwright-core/src/server/screenshotter.ts

```diff
@@ -58,29 +58,75 @@ function inPagePrepareForScreenshots(hideCaret: boolean, disableAnimations: bool
     return roots;
   };
 
-  let documentRoots: (Document|ShadowRoot)[] | undefined;
-  const memoizedRoots = () => documentRoots ??= collectRoots(document);
+  const roots = collectRoots(document);
+  const cleanupCallbacks: (() => void)[] = [];
 
-  const styleTags: Element[] = [];
-  if (hideCaret) {
-    for (const root of memoizedRoots()) {
-      const styleTag = document.createElement('style');
-      styleTag.textContent = `
-        *:not(#playwright-aaaaaaaaaa.playwright-bbbbbbbbbbb.playwright-cccccccccc.playwright-dddddddddd.playwright-eeeeeeeee) {
-          caret-color: transparent !important;
-        }
-      `;
-      if (root === document)
-        document.documentElement.append(styleTag);
-      else
-        root.append(styleTag);
-      styleTags.push(styleTag);
+  const hideCaretStyle = hideCaret ? `
+    *:not(#playwright-aaaaaaaaaa.playwright-bbbbbbbbbbb.playwright-cccccccccc.playwright-dddddddddd.playwright-eeeeeeeee) {
+      caret-color: transparent !important;
+    }` : '';
+
+  const styleText = `
+    ${hideCaretStyle}
+    [pw-mask-for-screenshot="true"] {
```

Comment:
Why separate attribute, and not `[data-test-screenshot="mask"]`?

Reply:
Because we only want it to be invisible during the screenshot.